README.md
# moment-business-days-sk

This is a momentJS plugin that allows you to work with only business days (Monday to Friday) except slovak holidays.

## Notes

- This plugin uses [momentjs-business-days](https://github.com/kalmecak/moment-business-days).
- This plugin sets the slovak locale (sk) format (DD/MM) and official annual slovak holidays.
- 30.10.2018 is one-time exceptional holiday [100. výročie prijatia Deklarácie slovenského národa](https://www.google.com/search?q=30.+oktober+2018+sviatok)
- This plugin computes Easter Sunday and adds Good Friday and Easter Monday to holiday
- Computed easter sundays are cached afterwards for later use
- It is possible to add additional, custom annual holidays (for local or district holidays) with slovak format (DD/MM).
- For the documentation see [momentjs-business-days](https://github.com/kalmecak/moment-business-days) and [moment](https://momentjs.com/docs/).
- Project [changelog](CHANGELOG.md)
- All contributions are welcome.

## Install

```bash
$ npm install --save moment-business-days-sk
```

## Usage TypeScript

```javascript
import { Moment } from 'moment';
import moment from 'moment-business-days-sk';

let easterSun = moment().easter(2020) as Moment; // Easter Sunday

let workingThu: Moment = moment(easterSun).prevBusinessDay(); // Thursday
let goodFri: Moment = moment(easterSun).subtract(2, 'd'); // Good Friday
let whiteSat: Moment = moment(easterSun).subtract(1, 'd'); // Easter Monday
let easterMon: Moment = moment(easterSun).add(1, 'd'); // Easter Monday
let workingTue: Moment = moment(easterSun).nextBusinessDay(); // Tuesday

console.log('%s isBusinessDay: %s', workingThu, workingThu.isBusinessDay());
console.log('%s isBusinessDay: %s', goodFri, goodFri.isBusinessDay());
console.log('%s isBusinessDay: %s', whiteSat, whiteSat.isBusinessDay());
console.log('%s isBusinessDay: %s', easterSun, easterSun.isBusinessDay());
console.log('%s isBusinessDay: %s', easterMon, easterMon.isBusinessDay());
console.log('%s isBusinessDay: %s', workingTue, workingTue.isBusinessDay());
```

## Usage JavaScript

```javascript
var moment = require('moment-business-days-sk');

console.log(moment('2020-10-01').isHoliday()); // false
console.log(moment().easter(2020)); // Moment<2020-04-12T00:00:00+02:00>
```

### Add a custom holiday

```javascript
var moment = require('moment-business-days-sk');

/*
moment-business-days-sk slovak holidays:
 holidays: [

    // Štátne sviatky
    '01/01', // 1. január – Deň vzniku Slovenskej republiky
    '05/07', // 5. júl – Sviatok svätého Cyrila a Metoda
    '29/08', // 29. august – Výročie Slovenského národného povstania
    '01/09', // 1. september – Deň Ústavy Slovenskej republiky
    '17/11', // 17. november – Deň boja za slobodu a demokraciu

    // Dni pracovného pokoja
    '06/01', // 6. január – Zjavenie Pána (Traja králi)
    '01/05', // 1. máj – Sviatok práce
    '08/05', // 8. máj – Deň víťazstva nad fašizmom
    '15/09', // 15. september – Sedembolestná Panna Mária
    '01/11', // 1. november – Sviatok všetkých svätých
    '24/12', // 24. december – Štedrý deň
    '25/12', // 25. december – Prvý sviatok vianočný
    '26/12', // 26. december – Druhý sviatok vianočný

    // Computed and cached
    // 'Easter Sunday'
    // Computed according to 'Easter Sunday'>
    //   'Good Friday'
    //   'Easter Monday'

    // One time exceptional case:
    '30/10/2018', // 100. výročie prijatia Deklarácie slovenského národa
 ]
*/
```

### Run Tests

`npm test`

## Methods to test slovak business days

**isHoliday**

```javascript
moment('2019-01-01').isHoliday(); // true
moment('2019-11-14').isHoliday(); // false
```

**easter**

```javascript
// 21.4.2019 is Easter Sunday
let easter = moment().easter(2019);

easter.month(); // 3 (4-1 = 3)
easter.date(); // 21

// Two days before each Easter is Friday and is holiday
let easterFriday = moment(easter).subtract(2, 'd');
easterFriday.day(); // 5 ... friday
easterFriday.isHoliday(); // true

// Next day after each Easter is Monday
let easterMonday = moment(easter).add(1, 'd');
easterMonday.day(); // 1 ... Monday
easterMonday.isHoliday(); // true
```

**isBusinessDay**

```javascript
// Year 2019 has 250 business days
let foundBusinessDays = 0;

for (let d = 0; d < 365; d++) {
let day = moment('2019-01-01').add(d, 'd');

if (day.isBusinessDay()) {
    foundBusinessDays++;
}
```

**prevBusinessDay, nextBusinessDay**

```javascript
let easter = moment().easter(2019);

let prevBusinessDay = moment(easter).prevBusinessDay();
prevBusinessDay.isHoliday(); // false
prevBusinessDay.day(); // 4 ... Thursday

let nextBusinessDay = moment(easter).nextBusinessDay();
nextBusinessDay.isHoliday(); // false
nextBusinessDay.day(); // 4 ... Tuesday
```

**lastBusinessDayOfThisMonth**

```javascript
let today = moment().startOf('day');

let firstDayOfNextMonth = moment(currDay).add(1, 'month').startOf('month');
let lastBusinessDayOfCurrMonth = moment(firstDayOfNextMonth).prevBusinessDay();
lastBusinessDayOfCurrMonth.isBusinessDay(); // true
```

## Methods from [moment-business-days](https://www.npmjs.com/package/moment-business-days)

**businessAdd(days)**

Will add just business days excluding Saturday and Sunday, return a moment date object:

```javascript
// 30-01-2015 is Friday, DD-MM-YYYY is the format
moment('30-01-2015', 'DD-MM-YYYY').businessAdd(3)._d; // Wed Feb 04 2015 00:00:00 GMT-0600 (CST)
```

**businessSubtract(days)**

Will subtract just business days excluding Saturday and Sunday, return a moment date object:

```javascript
// 27-01-2015 is Tuesday, DD-MM-YYYY is the format
moment('27-01-2015', 'DD-MM-YYYY').businessSubtract(3)._d; // Thu Jan 22 2015 00:00:00 GMT-0600 (CST)
```

**isBusinessDay()**

Check if the date is a business day and return **true**/**false**:

```javascript
// 31-01-2015 is Saturday
moment('31-01-2015', 'DD-MM-YYYY').isBusinessDay(); // false

// 30-01-2015 is Friday
moment('30-01-2015', 'DD-MM-YYYY').isBusinessDay(); // true
```

**nextBusinessDay()**

Will retrieve the next business date as moment date object:

```javascript
//Next busines day of Friday 30-01-2015
moment('30-01-2015', 'DD-MM-YYYY').nextBusinessDay()._d; // Mon Feb 02 2015 00:00:00 GMT-0600 (CST)

//Next busines day of Monday 02-02-2015
moment('02-02-2015', 'DD-MM-YYYY').nextBusinessDay()._d; //Tue Feb 03 2015 00:00:00 GMT-0600 (CST)
```

**prevBusinessDay()**

Will retrieve the previous business date as moment date object:

```javascript
//Previous busines day of Monday 02-02-2015
moment('02-02-2015', 'DD-MM-YYYY').prevBusinessDay()._d; // Fri Jan 30 2015 00:00:00 GMT-0600 (CST)

//Previous busines day of Tuesday 03-02-2015
moment('03-02-2015', 'DD-MM-YYYY').prevBusinessDay()._d; //Mon Feb 02 2015 00:00:00 GMT-0600 (CST)
```

**monthBusinessDays()**

Retrieve an array of the business days in the month, each one is a moment object.

```javascript
//Busines days in month January 2015
moment('01-01-2015', 'DD-MM-YYYY').monthBusinessDays();

/*
[ { _isAMomentObject: true,
    _i: '01-01-2015',
    _f: 'DD-MM-YYYY',
    _isUTC: false,
    _pf:{ ... },
    _locale: { ... },
    _d: Thu Jan 01 2015 00:00:00 GMT-0600 (CST)
  } {
   ...
  },
  ( ... )
]
*/
```

**monthNaturalDays()**

Is like monthBusinessDays(), but this method will include the weekends on it's response.

**monthBusinessWeeks()**

Retrieve an array of arrays, these arrays are the representation of a business weeks and each week (array) have it own business days (Monday to Friday). There could be the case that one week (array) have less than 5 days, this is because the month started on the middle of the week, for example: the first week of January 2015 just have two days, Thursday 1st and Friday 2nd. **Each day in the week arrays are moment objects.**

```javascript
//Busines weeks in month January 2015
moment('01-01-2015', 'DD-MM-YYYY').monthBusinessWeeks();

/*
[ [ { _isAMomentObject: true,
      _i: '01-01-2015',
      _f: 'DD-MM-YYYY',
      _isUTC: false,
      _pf: [...],
      _locale: [...],
      _d: Thu Jan 01 2015 00:00:00 GMT-0600 (CST) 
    }, { _isAMomentObject: true,
      _i: '01-01-2015',
      _f: 'DD-MM-YYYY',
      _isUTC: false,
      _pf: [...],
      _locale: [...],
      _d: Fri Jan 02 2015 00:00:00 GMT-0600 (CST) }
  ],
  [...]
]
*/
```

**monthNaturalWeeks()**

It's like monthBusinessWeeks(), but this method will include weekends on it's response.

The objects returned by functions are momentjs objects (**except isBusinessDay**) so you can handle it with moment native functions.

**businessDiff()**

Calculate number of busines days between dates.

```javascript
var diff = moment('05-15-2017', 'MM-DD-YYYY').businessDiff(moment('05-08-2017', 'MM-DD-YYYY'));
// diff = 5
```


CHANGELOG.md
# Changelog for 'moment-business-days-sk' package

All notable changes to this project will be documented in this file.
The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
GNU standard changelog format [GNU changelog style guide](https://www.gnu.org/prep/standards/html_node/Style-of-Change-Logs.html#Style-of-Change-Logs)
is not sufficient.

## [Unreleased]

- Work in progress

## [1.0.0] - 2020-10-01

- release

## [0.0.2] - 2020-01-31

### Added

- TypeScript
- Tests

## [0.0.1] - 2019-11-08

### Added

- The very first JavaScript version

[unreleased]: https://github.com/mirobartanus/moment-business-days-sk/compare/v1.0.0...HEAD
[1.0.0]: https://github.com/mirobartanus/moment-business-days-sk/compare/v0.0.2...v1.0.0
[0.0.2]: https://github.com/mirobartanus/moment-business-days-sk/compare/v0.0.1...v0.0.2
[0.0.1]: https://github.com/mirobartanus/moment-business-days-sk/releases/tag/v0.0.1

---

## How do I make a good changelog?

- Guiding Principles
- Changelogs are for humans, not machines.
- There should be an entry for every single version.
- The same types of changes should be grouped.
- Versions and sections should be linkable.
- The latest version comes first.
- The release date of each version is displayed.
- Mention whether you follow Semantic Versioning.

### Types of changes

- `Added` for new features.
- `Changed` for changes in existing functionality.
- `Deprecated` for soon-to-be removed features.
- `Removed` for now removed features.
- `Fixed` for any bug fixes.
- `Security` in case of vulnerabilities.

## [0.0.1] - YYYY-MM-DD

### Added

- a note about the added feature

### Changed

### Deprecated

### Removed

### Fixed

### Security


test/index.spec.ts
import { expect } from 'chai';
import 'mocha';

import { Moment } from 'moment';
import moment from '../src/index';

describe('moment-business-days-sk tests', () => {
  before(function () {
    moment.locale('sk');
  });

  it('Default value is "now", moment() === moment(new Date())', () => {
    const aDate: Date = new Date();
    const aMoment: Moment = moment();
    const aDiff = Math.abs(aMoment.valueOf() - aDate.getTime());

    expect(aDiff).to.be.lessThan(3); // 3x 1/1000 sec
  });

  it('default locale is "sk"', () => {
    expect(moment.locale()).to.be.eq('sk');
  });

  context('isHoliday', function () {
    it('isHoliday for 2019-01-01', () => {
      let day = moment('2019-01-01');
      expect(day.isHoliday()).to.equal(true);
    });

    it('isHoliday for 2019-11-14 - FALSE', () => {
      let day = moment('2019-11-14');
      expect(day.isHoliday()).to.equal(false);
    });

    it("isHoliday for Easter Friday '2019-04-19'", function () {
      let day = moment('2019-04-19');
      expect(day.isHoliday()).to.equal(true);
    });
  });

  context('easter', function () {
    it("returns [4,21] for '2019'", function () {
      let day = moment().easter(2019) as Moment;

      expect(day.month()).to.equal(4 - 1);
      expect(day.date()).to.equal(21);
    });

    it('500x Easter is on Sunday', function () {
      for (let y = 1750; y < 2250; y++) {
        let day = moment().easter(y) as Moment;

        expect(day.day()).to.equal(0);
      }
    });

    it('Second 500x Easter is much faster because caching', function () {
      for (let y = 1750; y < 2250; y++) {
        let day = moment().easter(y) as Moment;

        expect(day.day()).to.equal(0);
      }
    });

    it('Two days before each Easter is Friday', function () {
      for (let y = 1750; y < 2250; y++) {
        let easter = moment().easter(y) as Moment;
        let easterFriday = moment(easter).subtract(2, 'd');

        expect(easterFriday.day()).to.equal(5);
      }
    });

    it('Two days before each Easter is Friday which is Holiday', function () {
      for (let y = 1750; y < 2250; y++) {
        let easter = moment().easter(y) as Moment;
        let easterFriday = moment(easter).subtract(2, 'd');

        expect(easterFriday.isHoliday()).to.equal(true);
      }
    });

    it('Next day after each Easter is Monday', function () {
      for (let y = 1750; y < 2250; y++) {
        let easter = moment().easter(y) as Moment;
        let easterMonday = moment(easter).add(1, 'd');

        expect(easterMonday.day()).to.equal(1);
      }
    });

    it('Next day after each Easter is Monday which is Holiday', function () {
      for (let y = 1750; y < 2250; y++) {
        let easter = moment().easter(y) as Moment;
        let easterMonday = moment(easter).add(1, 'd');

        expect(easterMonday.isHoliday()).to.equal(true);
      }
    });

    it('Min and max months of Easter holiday are whitin 2-3', function () {
      let minMonth = 11;
      let maxMonth = 0;

      for (let y = 1750; y < 2250; y++) {
        let easter = moment().easter(y) as Moment;
        let easterFriday = moment(easter).subtract(2, 'd');
        let easterMonday = moment(easter).add(1, 'd');

        minMonth = Math.min(minMonth, easterFriday.month());
        maxMonth = Math.max(maxMonth, easterMonday.month());

        expect(minMonth).to.be.within(2, 3);
        expect(maxMonth).to.be.within(2, 3);
      }
    });
  });

  context('isHoliday', function () {
    it('Test all holidays in 2019', function () {
      let allHolidays = [
        '2019-01-01',
        '2019-01-06',
        '2019-04-19', // EASTER FRIDAY
        '2019-04-22', // EASTER MONDAY
        '2019-05-01',
        '2019-05-08',
        '2019-07-05',
        '2019-08-29',
        '2019-09-01',
        '2019-09-15',
        '2019-11-01',
        '2019-11-17',
        '2019-12-24',
        '2019-12-25',
        '2019-12-26',
      ];

      let foundHolidays = 0;

      for (let d = 0; d < 365; d++) {
        let day = moment('2019-01-01').add(d, 'd');

        if (day.isHoliday()) {
          foundHolidays++;
          expect(allHolidays.indexOf(day.format('YYYY-MM-DD')) >= 0).to.equal(true);
        }
      }

      expect(foundHolidays).to.equal(allHolidays.length);
    });

    it('Test all holidays in 1970-2070', function () {
      for (let y = 1970; y <= 2070; y++) {
        let yString = y.toString();

        let allHolidays = [
          yString + '-01-01',
          yString + '-01-06',
          //yString + '-04-19', // EASTER FRIDAY
          //yString + '-04-22', // EASTER MONDAY
          yString + '-05-01',
          yString + '-05-08',
          yString + '-07-05',
          yString + '-08-29',
          yString + '-09-01',
          yString + '-09-15',
          yString + '-11-01',
          yString + '-11-17',
          yString + '-12-24',
          yString + '-12-25',
          yString + '-12-26',
        ];

        let easterSunday = moment().easter(y) as Moment;

        allHolidays.push(easterSunday.subtract(2, 'd').format('YYYY-MM-DD'));
        allHolidays.push(easterSunday.add(3, 'd').format('YYYY-MM-DD'));

        if (y === 2018) {
          allHolidays.push('2018-10-30');
        }

        let foundHolidays = 0;

        for (let d = 0; d < 365; d++) {
          let day = moment(yString + '-01-01').add(d, 'd');

          if (day.isHoliday()) {
            foundHolidays++;

            expect(allHolidays.indexOf(day.format('YYYY-MM-DD')) >= 0).to.equal(true);
          }
        }

        expect(foundHolidays).to.equal(allHolidays.length);
      }
    });
  });

  context('isBusinessDay', function () {
    it('Rok 2019 ma 250 pracovnych dni', function () {
      let foundBusinessDays = 0;

      for (let d = 0; d < 365; d++) {
        let day = moment('2019-01-01').add(d, 'd');

        if (day.isBusinessDay()) {
          foundBusinessDays++;
        }
      }

      expect(foundBusinessDays).to.equal(250);
    });

    // sOURCE> https://calendar.zoznam.sk/worktime-sksk.php?hy=2021
    it('Test all business days in 2010-2030', function () {
      let yearBusinessDays = [
        { y: 2010, bd: 251 },
        { y: 2011, bd: 250 },
        { y: 2012, bd: 250 },
        { y: 2013, bd: 250 },
        { y: 2014, bd: 248 },
        { y: 2015, bd: 250 },
        { y: 2016, bd: 250 },
        { y: 2017, bd: 247 },
        { y: 2018, bd: 249 },
        { y: 2019, bd: 250 },
        { y: 2020, bd: 251 },
        { y: 2021, bd: 251 },
        { y: 2022, bd: 250 },
        { y: 2023, bd: 247 },
        { y: 2024, bd: 251 },
        { y: 2025, bd: 248 },
        { y: 2026, bd: 250 },
        { y: 2027, bd: 251 },
        { y: 2028, bd: 247 },
        { y: 2029, bd: 250 },
        { y: 2030, bd: 250 },
      ];

      for (let cYear of yearBusinessDays) {
        let startDateOfTheYear = moment([cYear.y]);
        let days = startDateOfTheYear.isLeapYear() ? 366 : 365;

        let foundBusinessDays = 0;

        for (let d = 0; d < days; d++) {
          let day = moment(startDateOfTheYear).add(d, 'd');

          if (day.isBusinessDay()) {
            foundBusinessDays++;
          }
        }

        expect(foundBusinessDays).to.equal(cYear.bd);
      }
    });
  });

  context('prevBusinessDay', function () {
    it('Predchadzajuci pracovny den pred Velkou Nocou 2019 je stvrtok, 18.4.2019', function () {
      let easter = moment().easter(2019) as Moment;
      let prevBusinessDay = moment(easter).prevBusinessDay();

      expect(prevBusinessDay.isHoliday()).to.equal(false);
      expect(prevBusinessDay.day()).to.equal(4);
      expect(prevBusinessDay.isSame(moment('2019-04-18'), 'day')).to.equal(true);
    });

    it('Predchadzajuci pracovny den pred Velkou Nocou je vzdy stvrtok', function () {
      for (let y = 1750; y < 2250; y++) {
        let easter = moment().easter(y) as Moment;
        let prevBusinessDay = moment(easter).prevBusinessDay();

        expect(prevBusinessDay.isHoliday()).to.equal(false);
        expect(prevBusinessDay.day()).to.equal(4);

        let thursday = moment(easter).subtract(3, 'd');
        expect(prevBusinessDay.isSame(thursday, 'day')).to.equal(true);
      }
    });
  });

  context('nextBusinessDay', function () {
    it('Nasledujuci pracovny den po Velkej Noci 2019 je utorok, 23.4.2019', function () {
      let easter = moment().easter(2019) as Moment;
      let nextBusinessDay = moment(easter).nextBusinessDay();

      expect(nextBusinessDay.isHoliday()).to.equal(false);
      expect(nextBusinessDay.day()).to.equal(2);
      expect(nextBusinessDay.isSame(moment('2019-04-23'), 'day')).to.equal(true);
    });

    it('Nasledujuci pracovny den po Velkej Noci je vzdy utorok', function () {
      for (let y = 1200; y < 2200; y++) {
        let easter = moment().easter(y) as Moment;
        let nextBusinessDay = moment(easter).nextBusinessDay();

        expect(nextBusinessDay.isHoliday()).to.equal(false);
        expect(nextBusinessDay.day()).to.equal(2);

        let tuesday = moment(easter).add(2, 'd');
        expect(nextBusinessDay.isSame(tuesday, 'day')).to.equal(true);
      }
    });
  });

  context('lastBusinessDayOfTheMonth', function () {
    it('Posledný pracovný deň v mesiaci', function () {
      for (let d = 0; d < 365; d++) {
        let currDay = moment('2019-01-01').startOf('d').add(d, 'd');

        let firstDayOfNextMonth = moment(currDay).startOf('d').add(1, 'month').startOf('month');

        let lastBusinessDayOfCurrMonth = moment(firstDayOfNextMonth).prevBusinessDay();
        expect(lastBusinessDayOfCurrMonth.isBusinessDay()).to.equal(true);

        let nextDay = moment(lastBusinessDayOfCurrMonth).add(1, 'd');
        while (!nextDay.isBusinessDay()) {
          nextDay.add(1, 'd');
        }

        expect(nextDay.isBusinessDay()).to.equal(true);
        expect(nextDay.month()).not.equal(lastBusinessDayOfCurrMonth.month());
      }
    });
  });

  context('addHoliday', function () {
    it('Custom holiday', function () {
      let notHoliday = moment('2020-10-01');

      expect(notHoliday.isBusinessDay()).to.equal(true);

      moment().addHoliday('01/10');

      expect(notHoliday.isBusinessDay()).to.equal(false);
    });
  });
});


src/index.ts
// https://www.typescriptlang.org/docs/handbook/declaration-merging.html
// https://www.typescriptlang.org/docs/handbook/namespaces-and-modules.html

//import moment, { Moment } from 'moment';
//import 'moment-business-days';

import { Moment } from 'moment';
import moment from 'moment-business-days';

declare module 'moment' {
  interface Moment {
    easter(year: number): moment.Moment | undefined;
    addHoliday(dateDDMM: string): void;
  }
}

const MIN_YEAR = 1000;
const MIN_YEAR_CACHED = 1750;
const MAX_YEAR_CACHED = 2250;
const MAX_YEAR = 3000;

moment.locale('sk');

moment.updateLocale('sk', {
  // https://sk.wikipedia.org/wiki/Zoznam_sviatkov_na_Slovensku
  // https://sk.wikipedia.org/wiki/De%C5%88_pracovn%C3%A9ho_pokoja
  holidays: [
    // Štátne sviatky
    '01/01', // 1. január – Deň vzniku Slovenskej republiky
    '05/07', // 5. júl – Sviatok svätého Cyrila a Metoda
    '29/08', // 29. august – Výročie Slovenského národného povstania
    '01/09', // 1. september – Deň Ústavy Slovenskej republiky
    '17/11', // 17. november – Deň boja za slobodu a demokraciu

    // Dni pracovného pokoja
    '06/01', // 6. január – Zjavenie Pána (Traja králi)
    '01/05', // 1. máj – Sviatok práce
    '08/05', // 8. máj – Deň víťazstva nad fašizmom
    '15/09', // 15. september – Sedembolestná Panna Mária
    '01/11', // 1. november – Sviatok všetkých svätých
    '24/12', // 24. december – Štedrý deň
    '25/12', // 25. december – Prvý sviatok vianočný
    '26/12', // 26. december – Druhý sviatok vianočný
  ],
  holidayFormat: 'DD/MM',
  holiday: function (someMoment: Moment) {
    // Month is indexed starting at 0, so December is 11.
    const year = someMoment.year();
    const month = someMoment.month() + 1;
    const date = someMoment.date();
    const day = someMoment.day();

    //  30. október 2018 je štátnym sviatkom: 100. výročie prijatia Deklarácie slovenského národa
    if (year === 2018 && month === 10 && 30 === date) {
      return true;
    }

    // on Fridays and Mondays durig March and April test Easter also:
    if ((month == 3 || month == 4) && (day == 5 || day == 1)) {
      const someEaster = someMoment.easter(year);

      if (day == 5) {
        const easterFriday = moment(someEaster).subtract(2, 'd');

        if (someMoment.isSame(easterFriday, 'day')) {
          return true;
        }
      }

      if (day == 1) {
        const easterMonday = moment(someEaster).add(1, 'd');

        if (someMoment.isSame(easterMonday, 'day')) {
          return true;
        }
      }
    }

    return false;
  },
});

// Veľkonočná nedeľa
//'2019': moment({ year: 2019, month: 4 - 1, day: 21 })
const easterCache: Map<number, Moment> = new Map<number, Moment>();

// https://github.com/zaygraveyard/moment-easter/blob/master/moment.easter.js
// https://stackoverflow.com/questions/1284314/easter-date-in-javascript
// https://www.irt.org/articles/js052/index.htm
function computeEaster(Y: number): { year: number; month: number; day: number } {
  const C = Math.floor(Y / 100);
  const N = Y - 19 * Math.floor(Y / 19);
  const K = Math.floor((C - 17) / 25);
  let I = C - Math.floor(C / 4) - Math.floor((C - K) / 3) + 19 * N + 15;
  I = I - 30 * Math.floor(I / 30);
  I = I - Math.floor(I / 28) * (1 - Math.floor(I / 28) * Math.floor(29 / (I + 1)) * Math.floor((21 - N) / 11));
  let J = Y + Math.floor(Y / 4) + I + 2 - C + Math.floor(C / 4);
  J = J - 7 * Math.floor(J / 7);
  const L = I - J;
  const M = 3 + Math.floor((L + 40) / 44);
  const D = L + 28 - 31 * Math.floor(M / 4);

  return { year: Y, month: M - 1, day: D };
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
function computeEasterAlt(year: number): { year: number; month: number; day: number } {
  /*jslint bitwise: true, vars: true */
  const a = ((year / 100) | 0) * 1483 - ((year / 400) | 0) * 2225 + 2613;
  const b = ((((year % 19) * 3510 + ((a / 25) | 0) * 319) / 330) | 0) % 29;
  // return 56 - b - ((year * 5 / 4 | 0) + a - b) % 7;
  const c = 148 - b - (((((year * 5) / 4) | 0) + a - b) % 7);

  return { year: year, month: ((c / 31) | 0) - 1, day: (c % 31) + 1 };
}

moment.fn.easter = function (year: number): Moment | undefined {
  if (!year) year = this.year();

  if (year < MIN_YEAR || year > MAX_YEAR) {
    return undefined;
  }

  let aEaster: Moment | undefined = undefined;

  if (year >= MIN_YEAR_CACHED && year <= MAX_YEAR_CACHED) {
    aEaster = easterCache.get(year);
  }

  if (aEaster === undefined) {
    aEaster = moment(computeEaster(year));

    if (year >= MIN_YEAR_CACHED && year <= MAX_YEAR_CACHED) {
      easterCache.set(year, aEaster as Moment);
    }
  }

  // return copy, to protect the value in chache
  return moment(aEaster);
};

moment.fn.addHoliday = function (dateDDMM: string) {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const locale: any = this.localeData();

  locale._holidays.push(dateDDMM);
};

export = moment;
